meta-balena-common/recipes-containers/hostapp-update/files/hostapp-update

```diff
@@ -37,19 +45,244 @@ run_current_hooks_and_recover () {
+#   SLUG
+#   target_supervisor_version
+# Arguments:
+#   image: the docker image to extract the config from
+# Returns:
+#   None
+#######################################
+upgrade_supervisor() {
+	local image=$1
+	INFO "Supervisor update start..."
+
+	if [ -z "$target_supervisor_version" ]; then
+		INFO "No explicit supervisor version was provided, update to default version in target balenaOS..."
+		local DEFAULT_SUPERVISOR_VERSION
+#		versioncheck_cmd=("run" "--rm" "${image}" "bash" "-c" "cat /etc/*-supervisor/supervisor.conf | sed -rn 's/SUPERVISOR_(TAG|VERSION)=v(.*)/\\2/p'")
+		DEFAULT_SUPERVISOR_VERSION=$(DOCKER_HOST="unix:///var/run/balena-host.sock" balena run --rm ${image} bash cat /etc/*-supervisor/supervisor.conf | sed -rn 's/SUPERVISOR_(TAG|VERSION)=v(.*)/\\2/p')
+		if [ -z "$DEFAULT_SUPERVISOR_VERSION" ]; then
+			ERROR "Could not get the default supervisor version for this balenaOS release, bailing out."
+		else
+			INFO "Extracted default version is v$DEFAULT_SUPERVISOR_VERSION..."
+			target_supervisor_version="$DEFAULT_SUPERVISOR_VERSION"
+		fi
+	fi
+
+	if supervisor_target_state_versions=$(_fetch_supervisor_version); then
+		echo ${supervisor_target_state_versions} | read -r CURRENT_SUPERVISOR_VERSION SCHEDULED_SUPERVISOR_VERSION
+		INFO "Supervisor state: Target ${target_supervisor_version}, current ${CURRENT_SUPERVISOR_VERSION}, scheduled ${SCHEDULED_SUPERVISOR_VERSION}"
+
+		# If scheduled higher than current and target, update to scheduled
+		# If scheduled not higher than current:
+		#  If target higher than current, patch and update to target
+		#  If target not higher than current, do nothing
+
+		if ! version_gt "${SCHEDULED_SUPERVISOR_VERSION}" "${CURRENT_SUPERVISOR_VERSION}"; then
+			# Supervisor target state current version is higher or equal than the scheduled version.
+			if version_gt "$target_supervisor_version" "$CURRENT_SUPERVISOR_VERSION" ; then
+				# Supervisor target version is higher than current target state version
+				INFO "Patching supervisor target state from v${CURRENT_SUPERVISOR_VERSION} to v${target_supervisor_version}"
+				progress 90 "Patching supervisor update"
+				if ! _patch_supervisor_version "$target_supervisor_version"; then
+				ERROR "Failed to patch supervisor version in target state, bailing out."
+				fi
+			else
+				INFO "Supervisor update: no update needed."
+				return 0
+			fi
+		else
+			# Supervisor target state scheduled version is higher than the current version
+			if version_gt "$SCHEDULED_SUPERVISOR_VERSION" "$target_supervisor_version" ; then
+				target_supervisor_version="$SCHEDULED_SUPERVISOR_VERSION"
+			fi
+		fi
+		INFO "Updating supervisor target state from v${CURRENT_SUPERVISOR_VERSION} to v${target_supervisor_version}"
+		progress 95 "Running supervisor update"
+		if ! _run_supervisor_update; then
+			WARN "Failed to update supervisor version - leave to next boot."
+		fi
+	else
+		ERROR "Failed to fetch current supervisor version from the API."
+	fi
+}
```

Comment:
Merging this into hostapp-update breaks the logical separation of hostapp and supervisor updates - I would move this logic to the update-balena-supervisor script as it has nothing to do with the hostapp.